src/main/kotlin/grpc/SetCookieInterceptor.kt

```diff
@@ -0,0 +1,97 @@
+
+private val logger = KotlinLogging.logger {}
+
+/**
+ * A [ServerInterceptor] implementation responsible for setting cookies in the gRPC response.
+ *
+ * This interceptor captures cookies that need to be set in the response and adds them to the trailers
+ * of the gRPC call when it is closed. It uses a context to store the cookies during the call processing.
+ */
+val setCookieInterceptor =
+    object : ServerInterceptor {
+        /**
+         * Returns a no-op listener. Used when the call chain cannot proceed.
+         */
+        private fun <ReqT> emptyListener(): ServerCall.Listener<ReqT?> = object : ServerCall.Listener<ReqT?>() {}
+
+        override fun <ReqT : Any?, RespT : Any?> interceptCall(
+            call: ServerCall<ReqT?, RespT?>?,
+            headers: Metadata?,
+            next: ServerCallHandler<ReqT?, RespT?>?,
+        ): ServerCall.Listener<ReqT?>? {
+            val cookiesToSet = mutableMapOf<String, String?>()
+            val context = Context.current().withValue(GrpcContext.COOKIES_TO_SET_CONTEXT_KEY, cookiesToSet)
+
+            val forwardingCall = createForwardingCall(call, cookiesToSet)
+
+            // Continue the call chain within the new context, so the service can access the cookiesToSet map
+            return context.call { next?.startCall(forwardingCall, headers) ?: emptyListener() }
+        }
+
+        /**
+         * Creates a forwarding call that sets cookies in the trailers when the call is closed.
+         *
+         * This function wraps the original gRPC call and intercepts the close method to add cookies
+         * to the trailers based on the context's cookiesToSet map. It ensures that cookies are only set
+         * if the call status is OK and there are cookies to set.
+         *
+         * @param ReqT The type of the request.
+         * @param RespT The type of the response.
+         * @param call The original gRPC call to forward.
+         * @param cookiesToSet A map of cookie names and values to be set in the response.
+         * @return A new [ServerCall] that forwards the original call and sets cookies in the trailers.
+         */
+        private fun <ReqT : Any?, RespT : Any?> createForwardingCall(
+            call: ServerCall<ReqT?, RespT?>?,
+            cookiesToSet: Map<String, String?>,
+        ): ServerCall<ReqT?, RespT?>? = object : SimpleForwardingServerCall<ReqT, RespT>(call) {
+            override fun close(status: Status, trailers: Metadata) {
+                if (status.isOk && cookiesToSet.isNotEmpty()) {
+                    logger.debug { "Setting cookies based on context: $cookiesToSet" }
+                    for ((name, value) in cookiesToSet) {
+                        val cookieString = buildCookieString(name, value)
+                        if (cookieString != null) {
+                            trailers.put(GrpcContext.COOKIE_METADATA_KEY, cookieString)
+                        }
+                    }
```

Comment:
(optional) beauty of functional programming:
```suggestion
                    cookiesToSet.entries
                        .mapNotNull { buildCookieString(it.key, it.value) }
                        .forEach { trailers.put(GrpcContext.COOKIE_METADATA_KEY, it) }
```

Reply:
Ohh, nice one!